Added Typechecking to input tensor in RNN

TamirCohen opened the pull request:
The input tensor of the RNN forward must be the same type as the weights.
While passing tensor of type long the error is:
    `RuntimeError: expected scalar type Long but found Float`
Which is misleading because it said to convert Something to Long, but the correct solution is to convert the input to Float (Which is the type of the weights).

The new error:
    `RuntimeError: input must have the type torch.float32, got type torch.int64`

Is correct and more verbose

Fixes #99998

TamirCohen committed:
Added Typechecking to input tensor in RNN

The input tensor must be the same type as the weights.
The old error:
    `RuntimeError: expected scalar type Long but found Float`
Is misleading because it said the convert Something to Long, but the correct solution is to convert the input to Float (Which is the type of the weights).

The new error:
    `RuntimeError: input must have the type torch.float32, got type torch.int64`

Is correct and more verbose

Skylion007 commented on torch/nn/modules/rnn.py:
```diff
@@ -209,6 +209,9 @@ def reset_parameters(self) -> None:
 
     def check_input(self, input: Tensor, batch_sizes: Optional[Tensor]) -> None:
+        if input.dtype != self._flat_weights[0].dtype:
+            raise RuntimeError('input must have the type {}, got type {}'.format(
```
Nit: this should be a ValueError, even if we don't consistently use that.

TamirCohen committed:
Changed error from RuntimeError to ValueError

TamirCohen commented:
@drisspg

drisspg reviewed (approved):
Looks good thanks!

TamirCohen committed:
Do not check type equallity if autocast is enables

The test failed because if torch.autocast is enabled the input and the weights can be with different types

mikaylagawarecki commented:
@pytorchbot merge
